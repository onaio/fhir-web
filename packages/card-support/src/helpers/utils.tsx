/** Function to download data to a file
 *
 * @param {string} data - data to be written to file
 * @param {string} filename - name of the file to be saved
 * @param {string} type - MIME type for the file
 */
export const downloadFile = (data: string, filename: string, type: string) => {
  const file = new Blob([data], { type });
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (window.navigator.msSaveOrOpenBlob) {
    // IE10+
    window.navigator.msSaveOrOpenBlob(file, filename);
  } else {
    // Others
    const docElement = document.createElement('a');
    const url = URL.createObjectURL(file);
    docElement.href = url;
    docElement.download = filename;
    document.body.appendChild(docElement);
    docElement.click();
    setTimeout(() => {
      document.body.removeChild(docElement);
      window.URL.revokeObjectURL(url);
    }, 0);
  }
};
