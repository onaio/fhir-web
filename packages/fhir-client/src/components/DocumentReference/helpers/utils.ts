import { get } from 'lodash';

/**
 * get object from an array, the check parameters are passed in as an obj, keys in keyValue can represent a path
 *
 * @param objArray - array obj with records of objs
 * @param keyValue - obj defining what key and values to use when matching
 * @param all - return first matched or all
 */
export function findObj<T extends object>(
  objArray?: T[],
  keyValue?: Record<keyof T | string, unknown>,
  all = false
) {
  if (objArray && keyValue) {
    const checkParameters = Object.entries(keyValue);
    const matched = objArray.filter((obj) => {
      let matches = true;
      checkParameters.forEach(([key, value]) => {
        if (get(obj, key) !== value) {
          matches = false;
        }
      });
      return matches;
    });
    if (all) {
      return matched;
    } else {
      return matched[0];
    }
  }
  if (!objArray || !keyValue) return;
}
