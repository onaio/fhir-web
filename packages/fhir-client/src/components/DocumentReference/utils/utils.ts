import { get, isEmpty, some } from 'lodash';
import { IfhirR4 } from '@smile-cdr/fhirts';
import { format } from 'date-fns';
import { dateFormat } from '../../../constants';
import FHIR from 'fhirclient';

/**
 * checks if collection is empty or if all values in the obj can be said to hold null values
 *
 * @param coll - the collection
 */
export const itemIsEmpty = <T extends object>(coll?: T) => {
  if (isEmpty(coll)) {
    return true;
  }
  return !some(coll);
};

/**
 * process top level fields of the documentReference resource
 *
 * @param docResource DocumentResource
 */
export const processTopLevelFields = (docResource: IfhirR4.IDocumentReference) => {
  const { status, type, description, securityLabel, date, id } = docResource;
  return {
    id,
    title: description,
    createdAt: date ? format(new Date(date), dateFormat) : undefined,
    status,
    documentType: {
      codeList: type?.coding,
      code: get(type, 'coding.0'),
    },
    securityCodes: {
      codeList: securityLabel,
      code: get(securityLabel, '[0].coding[0]'),
    },
  };
};

/**
 * Process context field in a documentReference
 *
 * @param docResource - the document resource
 */
export const processContextFields = (docResource: IfhirR4.IDocumentReference) => {
  const start = get(docResource, 'context.period.start');
  const end = get(docResource, 'context.period.end');
  return {
    periodStart: start ? format(new Date(start), dateFormat) : undefined,
    periodEnd: end ? format(new Date(end), dateFormat) : undefined,
    eventCoding: get(docResource, 'context.event[0].coding[0]'),
    facilityTypeCoding: get(docResource, 'context.facilityType.coding[0]'),
    practiceSettingCoding: get(docResource, 'context.practiceSetting.coding[0]'),
  };
};

/**
 * extract values from content field, will include the attachments and their formats
 *
 * @param docResource - the document resource
 */
export const processContentFields = (docResource: IfhirR4.IDocumentReference) => {
  return get(docResource, 'content').map((content) => {
    const data = get(content, 'attachment.data');
    const url = get(content, 'attachment.url');
    const absoluteUrlRegex = /^(?:[a-z]+:)?\/\//;
    const binaryUrlRegex = /Binary\/[\w-]+$/gm;
    let urlIsAbsolute, binaryUrl;
    if (url) {
      urlIsAbsolute = absoluteUrlRegex.test(url);
      const matches = Array.from(
        url.matchAll(binaryUrlRegex),
        (m) => (m as Array<string | undefined>)[0]
      );
      binaryUrl = matches[0];
    }
    const size = get(content, 'attachment.size');
    const formatCode = get(content, 'format.code');
    const formatDisplay = get(content, 'format.display');
    const formatContentType = get(content, 'attachment.contentType');

    return {
      attachment: { url, urlIsAbsolute, binaryUrl, data },
      size,
      formatCode,
      formatDisplay,
      formatContentType,
    };
  });
};

/**
 * parses docResources and extracts attachment data, filtering out those with empty string data uris
 *
 * @param docResources - documentReference resource array
 */
export const processDocumentReferences = (docResources: IfhirR4.IDocumentReference[]) => {
  return docResources.map((resource) => {
    return {
      ...processTopLevelFields(resource),
      context: processContextFields(resource),
      content: processContentFields(resource),
    };
  });
};

export type ParsedDocReference = ReturnType<typeof processDocumentReferences>;

/**
 * get the discrete and multipart of a content-type
 *
 * @param contentType - mime type
 */
export const splitContentType = (contentType: string) => {
  const splitText = contentType.split('/');
  return { discretePart: splitText[0], multipart: splitText[1] };
};

/**
 * Act as a proxy when downloading attachments so that we can customize the name.
 * This pulls attachment as blob from fhir /Binary endpoint
 *
 * @param fhirBaseURL -the fhir base url
 * @param binaryEndpoint - a Binary endpoint url path
 */
export const fetchAttachmentForDownload = async (fhirBaseURL: string, binaryEndpoint: string) => {
  return FHIR.client(fhirBaseURL)
    .request({
      url: binaryEndpoint,
    })
    .then((res) => {
      return res.blob();
    });
};
