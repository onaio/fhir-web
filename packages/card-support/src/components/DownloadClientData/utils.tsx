import Papaparse from 'papaparse';
import { OpenSRPService } from '@opensrp/server-service';
import { DownloadClientDataFormFields } from '../DownloadClientData';
import { Dispatch, SetStateAction } from 'react';
import { OPENSRP_URL_CLIENT_SEARCH, APPLICATION_CSV } from '../../constants';
import { Client } from '../../ducks/clients';
import { downloadFile } from '../../helpers/utils';
import { ParsedHierarchyNode } from '@opensrp/location-management';
import { sendErrorNotification } from '@opensrp/notifications';
import { Dictionary } from '@onaio/utils';
import lang, { Lang } from '../../lang';
/* eslint-disable @typescript-eslint/camelcase */

/** interface for user assignment response */
export interface UserAssignment {
  organizationIds: number[];
  jurisdictions: string[];
  plans: string[];
}

/** interface for CSV entries */
export interface ClientCSVEntry {
  id: string;
  dob: string;
  first_name: string;
  last_name: string;
  gender: string;
  facility_of_registration: string;
  date_of_registration: string;
  card_status: string;
  card_status_last_update: string;
}

/**
 * Handle form date change
 *
 * @param {string} range - selected date range
 * @param {Dispatch<SetStateAction<[string, string]>>} setCardOrderDate - state dispatcher
 */
export const handleCardOrderDateChange = (
  range: [string, string],
  setCardOrderDate: Dispatch<SetStateAction<[string, string]>>
): void => {
  setCardOrderDate(range);
};

/**
 * Handle download client data form submission
 *
 * @param values - sumitted form values
 * @param accessToken - OPENSRP API access token
 * @param opensrpBaseURL - OPENSRP API base URL
 * @param serviceClass - OPENSRP service class
 * @param locations location hierarchy
 * @param setSubmitting - method to set form `isSubmitting` status
 * @param langObj - tthe lang object
 */
export const submitForm = (
  values: DownloadClientDataFormFields,
  accessToken: string,
  opensrpBaseURL: string,
  serviceClass: typeof OpenSRPService,
  locations: ParsedHierarchyNode[],
  setSubmitting: (isSubmitting: boolean) => void,
  langObj: Lang = lang
): void => {
  const { clientLocation, cardStatus, cardOrderDate } = values;

  if (!clientLocation) {
    return;
  }

  setSubmitting(true);
  const startDate = cardOrderDate[0];
  const endDate = cardOrderDate[1];
  const endPoint = `${OPENSRP_URL_CLIENT_SEARCH}`;
  const serve = new serviceClass(accessToken, opensrpBaseURL, endPoint);
  let params: Dictionary = {
    locationIds: clientLocation,
  };

  if (cardStatus) {
    params = {
      ...params,
      attribute: `card_status:${cardStatus}`,
    };
  }

  serve
    .list(params)
    .then((clients: Client[]) => {
      const location = getLocationDetails(locations, clientLocation);
      const entries: ClientCSVEntry[] = clients
        .filter((client: Client) => !!client.identifiers.zeir_id)
        .filter((client: Client) => {
          const dateCreated = new Date(client.dateCreated.split('T')[0]);

          return (
            dateCreated.getTime() >= new Date(startDate).getTime() &&
            dateCreated.getTime() <= new Date(endDate).getTime()
          );
        })
        .map((client: Client) => {
          return {
            id: client.identifiers.zeir_id,
            dob: formatDDMMYYY(new Date(client.birthdate.split('T')[0])),
            first_name: client.firstName,
            last_name: client.lastName,
            gender: client.gender,
            facility_of_registration: location ? location.title : '',
            date_of_registration: formatDDMMYYY(new Date(client.dateCreated.split('T')[0])),
            card_status: client.attributes.card_status,
            card_status_last_update: formatDDMMYYY(
              new Date(client.attributes.card_status_date.split('T')[0])
            ),
          };
        });

      setSubmitting(false);

      if (entries.length) {
        createCsv(entries, buildCSVFileName(location ? location.title : '', startDate, endDate));
      } else {
        sendErrorNotification(langObj.NO_DATA_FOUND);
      }
    })
    .catch((_: Error) => {
      setSubmitting(false);
      sendErrorNotification(langObj.ERROR_OCCURRED);
    });
};

/**
 * Format a Date object as dd MM YYYY
 *
 * @param {Date} date - date object
 * @param {string} delimiter - delimiter to use to separate the different parts
 * @returns {string} formatted date
 */
export const formatDDMMYYY = (date: Date, delimiter = '/'): string => {
  //eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  return `${('0' + date.getDate()).slice(-2)}${delimiter}${('0' + (date.getMonth() + 1)).slice(
    -2
  )}${delimiter}${date.getFullYear()}`;
};

/**
 * Build the CSV file name
 *
 * @param {string} location - name of the location filter
 * @param {string} startDate - start date of the range filter
 * @param {string} endDate - end date of the range filter
 * @returns {string} CSV file name
 */
export const buildCSVFileName = (location: string, startDate: string, endDate: string): string => {
  return `Children_list_${location}_${formatDDMMYYY(new Date(), '_')}_(${formatDDMMYYY(
    new Date(startDate),
    '-'
  )} - ${formatDDMMYYY(new Date(endDate), '-')})`;
};

// Create csv data for the selected jurisdiction hieracrhy
export const createCsv = (entries: ClientCSVEntry[], fileName: string): void => {
  const csv: string = Papaparse.unparse(entries, {
    header: true,
  });
  // Export csv file
  downloadFile(csv, `${fileName}.csv`, APPLICATION_CSV);
};

/**
 * Get location from the location hierarchy
 *
 * @param {ParsedHierarchyNode} locations location hierarchy
 * @param {string} locationId id of location to return
 * @returns {ParsedHierarchyNode | null} location if found, null if not found
 */
export const getLocationDetails = (
  locations: ParsedHierarchyNode[],
  locationId: string
): ParsedHierarchyNode | null => {
  let found = false;
  let i = 0;
  let location = null;

  while (!found && i < locations.length) {
    const node = locations[i];

    if (node.id && node.id === locationId) {
      found = true;
      location = node;
    } else if (node.children) {
      location = getLocationDetails(node.children, locationId);

      if (location) {
        found = true;
      }
    }

    i += 1;
  }

  return location;
};
