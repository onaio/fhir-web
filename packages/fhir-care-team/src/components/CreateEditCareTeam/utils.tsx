import { history } from '@onaio/connected-reducer-registry';
import { v4 } from 'uuid';
import FHIR from 'fhirclient';
import { sendErrorNotification, sendSuccessNotification } from '@opensrp/notifications';
import lang from '../../lang';
import { URL_CARE_TEAM } from '../../constants';
import { Dictionary } from '@onaio/utils';
import { IfhirR4 } from '@smile-cdr/fhirts';
import { FormFields } from './Form';

export const submitForm = async (
  values: FormFields,
  fhirBaseURL: string,
  id?: string,
  uuid?: string
): Promise<void> => {
  const careTeamId = uuid ? uuid : v4();
  const payload: Omit<IfhirR4.ICareTeam, 'meta'> = {
    resourceType: 'CareTeam',
    identifier: [
      {
        use: 'official',
        value: careTeamId,
      },
    ],
    id: id,
    name: values.name,
    status: values.status as IfhirR4.CareTeam.StatusEnum,
    subject: {
      reference: `Group/${values.groupsId}`,
    },
    participant: values.practitionersId?.map((id) => {
      return {
        member: {
          reference: `Practitioner/${id}`,
        },
      };
    }),
  };
  const serve = FHIR.client(fhirBaseURL);
  if (id) {
    await serve
      .update(payload)
      .then(() => sendSuccessNotification(lang.CARE_TEAMS_UPDATE_SUCCESS))
      .catch(() => sendErrorNotification(lang.ERROR_OCCURRED));
  } else {
    await serve
      .create(payload)
      .then(() => sendSuccessNotification(lang.CARE_TEAMS_ADD_SUCCESS))
      .catch(() => sendErrorNotification(lang.ERROR_OCCURRED));
  }
  history.push(URL_CARE_TEAM);
};

/** Util function to build out patient or practitioner name
 *
 * @param {object} patient - patient resource object
 * @returns {string} - returns patient or practitioner name string
 */
export function getPatientName(patient: Dictionary | undefined) {
  if (!patient) {
    return '';
  }

  let name = patient.name;
  if (!Array.isArray(name)) {
    name = [name];
  }
  name = name[0];
  if (!name) {
    return '';
  }

  const family = Array.isArray(name.family) ? name.family : [name.family];
  const given = Array.isArray(name.given) ? name.given : [name.given];
  const prefix = Array.isArray(name.prefix) ? name.prefix : [name.prefix];
  const suffix = Array.isArray(name.suffix) ? name.suffix : [name.suffix];

  return [
    prefix.map((t = '') => String(t).trim()).join(' '),
    given.map((t = '') => String(t).trim()).join(' '),
    family.map((t = '') => String(t).trim()).join(' '),
    suffix.map((t = '') => String(t).trim()).join(' '),
  ]
    .filter(Boolean)
    .join(' ');
}
