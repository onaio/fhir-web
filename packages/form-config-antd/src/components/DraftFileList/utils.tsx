import React from 'react';
import { Dictionary } from '@onaio/utils';
import { ManifestFilesTypes, formatDate } from '@opensrp/form-config-core';
import { TableActions } from './TableActions';
import { getFetchOptions } from '@opensrp/server-service';

/**
 * Return table columns
 *
 * @param {string} accessToken Opensrp API access token
 * @param {string} opensrpBaseURL Opensrp API base URL
 * @param {boolean} isJsonValidator true if is json validator, false otherwise
 * @param {Dictionary} sortedInfo object containing sort order information
 * @param {Function} customFetchOptions Opensrp API custom options
 * @returns {Dictionary[]} table columns
 */
export const getTableColumns = (
  accessToken: string,
  opensrpBaseURL: string,
  isJsonValidator: boolean,
  sortedInfo?: Dictionary,
  customFetchOptions?: typeof getFetchOptions
): Dictionary[] => {
  const columns: Dictionary[] = [];
  const headerItems: string[] = ['Identifier', 'File Name', 'File Version', 'Created At', 'Module'];
  const fields: string[] = ['identifier', 'label', 'version', 'createdAt', 'module'];

  fields.forEach((field: string, index: number) => {
    let column: Dictionary = {
      title: headerItems[index],
      dataIndex: fields[index],
      key: fields[index],
      sorter: (a: Dictionary, b: Dictionary) => {
        return a[fields[index]].length - b[fields[index]].length;
      },
      sortOrder: sortedInfo && sortedInfo.columnKey === fields[index] && sortedInfo.order,
      ellipsis: true,
    };

    if (field === 'createdAt') {
      column = {
        ...column,
        render: (value: string) => formatDate(value),
      };
    }

    columns.push(column);
  });

  columns.push({
    title: 'Action',
    key: 'action',
    // eslint-disable-next-line react/display-name
    render: (_: string, file: ManifestFilesTypes) => {
      const tableActionProps = {
        file,
        accessToken,
        opensrpBaseURL,
        isJsonValidator,
        customFetchOptions,
      };
      return <TableActions {...tableActionProps} />;
    },
  });
  return columns;
};
