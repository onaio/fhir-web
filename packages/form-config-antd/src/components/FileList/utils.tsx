import { Dictionary } from '@onaio/utils';
import { formatDate } from '@opensrp/form-config-core';
import lang, { Lang } from '../../lang';

/**
 * Return table columns
 *
 * @param isJsonValidator boolean to check whether is Json validator
 * @param sortedInfo object containing sort order information
 * @param langObj the language translation object
 * @returns {Dictionary[]} table columns
 */
export const getTableColumns = (
  isJsonValidator: boolean,
  sortedInfo?: Dictionary,
  langObj: Lang = lang
): Dictionary[] => {
  const columns: Dictionary[] = [];
  const headerItems: string[] = [
    langObj.IDENTIFIER,
    langObj.FILE_NAME,
    langObj.FILE_VERSION,
    langObj.CREATED_AT,
  ];
  const fields: string[] = ['identifier', 'label', 'version', 'createdAt'];

  fields.forEach((field: string, index: number) => {
    let column: Dictionary = {
      title: headerItems[index],
      dataIndex: fields[index],
      key: fields[index],
      sorter: (a: Dictionary, b: Dictionary) => {
        return a[fields[index]].length - b[fields[index]].length;
      },
      sortOrder: sortedInfo && sortedInfo.columnKey === fields[index] && sortedInfo.order,
      ellipsis: true,
    };

    if (field === 'createdAt') {
      column = {
        ...column,
        render: (value: string) => formatDate(value),
      };
    }

    columns.push(column);
  });

  if (!isJsonValidator) {
    columns.push({
      title: langObj.MODULE,
      dataIndex: 'module',
      key: 'module',
    });
  }

  return columns;
};
