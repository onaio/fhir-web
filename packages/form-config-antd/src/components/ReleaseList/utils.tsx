import { Dictionary } from '@onaio/utils';
import { formatDate } from '@opensrp/form-config-core';

/**
 * Return table columns
 *
 * @param {Dictionary} sortedInfo object containing sorting order information
 * @returns {Dictionary[]} table columns
 */
export const getTableColumns = (sortedInfo?: Dictionary): Dictionary[] => {
  const columns: Dictionary[] = [];
  const headerItems: string[] = ['Identifier', 'App Id', 'App Version', 'Updated At'];
  const fields: string[] = ['identifier', 'appId', 'appVersion', 'updatedAt'];

  fields.forEach((field: string, index: number) => {
    let column: Dictionary = {
      title: headerItems[index],
      dataIndex: fields[index],
      key: fields[index],
      sorter: (a: Dictionary, b: Dictionary) => {
        return a[fields[index]].length - b[fields[index]].length;
      },
      sortOrder: sortedInfo && sortedInfo.columnKey === fields[index] && sortedInfo.order,
      ellipsis: true,
    };

    if (field === 'identifier' || field === 'appVersion') {
      column = {
        ...column,
        render: (value: string) => `V${value}`,
      };
    }

    if (field === 'updatedAt') {
      column = {
        ...column,
        render: (value: string) => formatDate(value),
      };
    }

    columns.push(column);
  });

  return columns;
};
