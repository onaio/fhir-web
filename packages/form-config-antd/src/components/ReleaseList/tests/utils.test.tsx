import { getTableColumns } from '../utils';

describe('components/Antd/ReleaseList/utils/getTableColumns', () => {
  const viewReleaseURL = '/releases';

  it('returns table columns', () => {
    expect(getTableColumns(viewReleaseURL)).toMatchSnapshot('table columns');
  });
});
