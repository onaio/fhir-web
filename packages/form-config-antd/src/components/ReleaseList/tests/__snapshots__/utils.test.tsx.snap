// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Antd/ReleaseList/utils/getTableColumns returns table columns: table columns 1`] = `
Array [
  Object {
    "dataIndex": "identifier",
    "ellipsis": true,
    "key": "identifier",
    "render": [Function],
    "sortOrder": false,
    "sorter": [Function],
    "title": "Identifier",
  },
  Object {
    "dataIndex": "appId",
    "ellipsis": true,
    "key": "appId",
    "sortOrder": false,
    "sorter": [Function],
    "title": "App Id",
  },
  Object {
    "dataIndex": "appVersion",
    "ellipsis": true,
    "key": "appVersion",
    "render": [Function],
    "sortOrder": false,
    "sorter": [Function],
    "title": "App Version",
  },
  Object {
    "dataIndex": "updatedAt",
    "ellipsis": true,
    "key": "updatedAt",
    "render": [Function],
    "sortOrder": false,
    "sorter": [Function],
    "title": "Updated At",
  },
]
`;
